minimal_scheduling/minimal_scheduling_real_time_tutorial.cpp

```diff
@@ -0,0 +1,153 @@
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count1_++);
+        this->publisher1_->publish(message);
+      };
+    timer1_ = this->create_wall_timer(500ms, timer_callback1);
+
+    publisher2_ = this->create_publisher<std_msgs::msg::String>("topic_rt", 10);
+    auto timer_callback2 =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count2_++);
+        this->publisher2_->publish(message);
+      };
+    timer2_ = this->create_wall_timer(500ms, timer_callback2);
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer1_, timer2_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher1_, publisher2_;
+  size_t count1_, count2_;
+};
+
+class MinimalSubscriber : public rclcpp::Node
+{
+public:
+  MinimalSubscriber(const std::string & node_name, const std::string & topic_name)
+  : Node(node_name), context_switches_counter_(RUSAGE_THREAD)
+  {
+    subscription1_ = this->create_subscription<std_msgs::msg::String>(
+      topic_name,
+      10,
+      [this](std_msgs::msg::String::UniquePtr) {
+        auto context_switches = context_switches_counter_.get();
+        if (context_switches > 0L) {
+          RCLCPP_WARN(
+            this->get_logger(), "[sub]    Involuntary context switches: '%lu'",
+            context_switches);
+        } else {
+          RCLCPP_INFO(
+            this->get_logger(), "[sub]    Involuntary context switches: '%lu'",
+            context_switches);
+        }
+        burn_cpu_cycles(200ms);
+      });
+  }
+
+private:
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription1_;
+  ContextSwitchesCounter context_switches_counter_;
+};
+
+
+int main(int argc, char * argv[])
+{
+  // Force flush of the stdout buffer.
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  auto options_reader = SchedOptionsReader();
+  if (!options_reader.read_options(argc, argv)) {
+    options_reader.print_usage();
+    return 0;
+  }
+  auto options = options_reader.get_options();
+
+  rclcpp::init(argc, argv);
+
+  auto node_pub = std::make_shared<MinimalPublisher>();
+  auto node_sub = std::make_shared<MinimalSubscriber>("minimal_sub1", "topic");
+  auto node_sub_rt = std::make_shared<MinimalSubscriber>("minimal_sub2", "topic_rt");
+
+  rclcpp::executors::StaticSingleThreadedExecutor default_executor;
+  rclcpp::executors::StaticSingleThreadedExecutor realtime_executor;
+
+  // the publisher and non real-time subscriber are processed by default_executor
+  default_executor.add_node(node_pub);
+  default_executor.add_node(node_sub);
+
+  // real-time subscriber is processed by realtime_executor.
+  realtime_executor.add_node(node_sub_rt);
+
+  // middleware threads are not configured with real-time priority
+  // because scheduling configuration of main thread is not real-time.
+
+  // Note: you can only add complete nodes to an Executor. That is:
```

Comment:
@JanStaschulat It is possible to only a subset of subscriptions to an executor if they are excluded from the default callback group. There is an example here: https://github.com/ros-realtime/ros2-realtime-examples/blob/c09aa1f26b249b21366288f76986d7fd13e7a808/minimal_scheduling/minimal_scheduling_callback_group.cpp

I'm fine showing mixed criticality using both options:
- Using callback groups
- Using separate nodes

Reply:
Yes, mixed-criticality can be achieved with different functionality. With this example I wanted to show the case with only two threads and two executors. 

In a second tutorial, we can describe mixed criticality with your existing example with callback groups. I thought, this is a bit more advanced, so I wanted to have a simpler code example as first option.